processor/src/advice/mod.rs

```diff
@@ -7,188 +7,113 @@ use vm_core::{
+    ///
+    /// Note: a word is always stored as little-endian. A `[...,a,b,c,d]` tape will yield
+    /// `[d,c,b,a]`.
     ///
     /// # Errors
     /// Returns an error if the advice tape does not contain a full word.
-    pub fn read_tapew(&mut self) -> Result<Word, ExecutionError> {
-        if self.tape.len() < 4 {
-            return Err(ExecutionError::AdviceTapeReadFailed(self.step));
-        }
-
-        let idx = self.tape.len() - 4;
-        let result = [self.tape[idx + 3], self.tape[idx + 2], self.tape[idx + 1], self.tape[idx]];
-
-        self.tape.truncate(idx);
-
-        Ok(result)
-    }
+    fn read_tape_w(&mut self) -> Result<Word, ExecutionError>;
 
-    /// Removes the next two words from the advice tape and returns them.
+    /// Pops a double word (8 elements) from the advice tape and returns them.
+    ///
+    /// Note: a double word is always stored as little-endian. A `[...,a,b,c,d,e,f,g,h]` tape will
+    /// yield `[h,g,f,e,d,c,b,a]`.
     ///
     /// # Errors
     /// Returns an error if the advice tape does not contain two words.
-    pub fn read_tape_double(&mut self) -> Result<[Word; 2], ExecutionError> {
-        let word0 = self.read_tapew()?;
-        let word1 = self.read_tapew()?;
-
-        Ok([word0, word1])
-    }
+    fn read_tape_dw(&mut self) -> Result<[Word; 2], ExecutionError>;
 
     /// Writes the provided value at the head of the advice tape.
-    pub fn write_tape(&mut self, value: Felt) {
-        self.tape.push(value);
-    }
+    fn write_tape(&mut self, value: Felt);
 
-    /// Retrieves a list of elements from a key-value map for the specified key, reverses it, and
-    /// writes the reversed list at the head of the advice tape. This way, the first element in the
-    /// list is located at the head of the advice tape.
+    /// Fetch a keyed tape from the values map, reversing and appending it to the advice tape.
+    ///
+    /// Note: this operation shouldn't consume the map element so it can be called multiple times
+    /// for the same key.
+    ///
+    /// # Example
+    /// Given an advice map `[a,b,c]`, and a map `x |-> [d,e,f]`, a call `write_tape_from_map(x)`
+    /// will result in `[a,b,c,f,e,d]` for the advice tape, and will preserve `x |-> [d,e,f]`.
     ///
     /// # Errors
     /// Returns an error if the key was not found in a key-value map.
-    pub fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError> {
-        let values = self
-            .values
-            .get(&key.into_bytes())
-            .ok_or(ExecutionError::AdviceKeyNotFound(key))?;
-        for &elem in values.iter().rev() {
-            self.tape.push(elem);
-        }
-
-        Ok(())
-    }
-
-    /// Inserts a list of elements to the advice map with the top four elements of the stack as
-    /// the key.
+    fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError>;
+
+    /// Maps a key to a values list to be yielded by `write_tape_from_map`.
```

Comment:
nit: probably should be "value list" rather than "values list".

Reply:
True, done!